server/src/main/java/io/druid/server/lookup/cache/LookupCoordinatorManager.java

```diff
@@ -506,137 +274,426 @@ public boolean deleteLookup(final String tier, final String lookup, AuditInfo au
+          LOG.warn("Background lookup manager thread could not be cancelled");
         }
 
-        @Override
-        public void onFailure(Throwable t)
-        {
-          backgroundManagerExitedLatch.countDown();
-          if (backgroundManagerFuture.isCancelled()) {
-            LOG.info("Background lookup manager exited");
-            LOG.trace(t, "Background lookup manager exited with throwable");
-          } else {
-            LOG.makeAlert(t, "Background lookup manager exited with error!").emit();
+        if (executorService != null) {
+          executorService.shutdownNow();
+        }
+
+        LOG.debug("Stopped");
+      }
+      catch (Exception ex) {
+        LOG.makeAlert(ex, "Got Exception while stop()").emit();
+      }
+      finally {
+        //so that next start() would happen.
+        lifecycleLock.exitStop();
+        lifecycleLock.reset();
+      }
+    }
+  }
+
+  private void lookupMgmtLoop()
+  {
+    // Sanity check for if we are shutting down
+    if (Thread.currentThread().isInterrupted() || !lifecycleLock.awaitStarted(1, TimeUnit.MILLISECONDS)) {
+      LOG.info("Not updating lookups because process was interrupted or not finished starting yet.");
+      return;
+    }
+
+    final Map<String, Map<String, LookupExtractorFactoryMapContainer>> allLookupTiers = lookupMapConfigRef.get();
+
+    if (allLookupTiers == null) {
+
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Not updating lookups because no data exists");
+      }
+      return;
+    }
+
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Starting lookup sync for on all nodes.");
+    }
+
+    final Map<HostAndPort, LookupsStateWithMap> currState = new HashMap<>();
+    final Random rand = new Random();
+
+    try {
+      List<ListenableFuture<?>> futures = new ArrayList<>();
+      for (String tier : allLookupTiers.keySet()) {
+
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("Starting lookup mgmt for tier [%s].", tier);
+        }
+
+        final Map<String, LookupExtractorFactoryMapContainer> tierLookups = allLookupTiers.get(tier);
+        for (final HostAndPort node : listenerDiscoverer.getNodes(LookupModule.getTierListenerPath(tier))) {
+
+          if (LOG.isDebugEnabled()) {
+            LOG.debug("Starting lookup mgmt for tier [%s] and host [%s:%s].", tier, node.getHostText(), node.getPort());
           }
+
+          futures.add(
+              executorService.submit(
+                  new Runnable()
+                  {
+                    @Override
+                    public void run()
+                    {
+                      try {
+
+                        if (LOG.isDebugEnabled()) {
+                          LOG.debug("Starting lookup sync for node [%s].", node);
+                        }
+
+                        LookupsStateWithMap lookupsState = knownOldState.get() != null ? knownOldState.get().get(node) : null;
+                        if (lookupsState == null
```

Comment:
removed the condition.... i was trying to optimize a bit by looking at previous known state of the node... if it had nothing in toLoad/toDrop , then I was assuming that nothing would have changed on that node ... but still added the random part to refresh just in case... when looking at the code again after all this while... i believe that optimization is not necessary and it is ok to refresh node state every time